File: game/state.py
from typing import Optional, Type

import global_state
from game_data import GameData
from phone_events import PhoneEvents

class State:
    def __init__(self, context: GameData):
        self.start_time = None
        self.events = {}
        self.context = context

    def process_events(self, phone_events: PhoneEvents) -> Optional[Type['State']]:
        # Do nothing
        pass

    def render(self, screen):
        # Do nothing
        pass

    def on_enter(self) -> None:
        self.start_time = global_state.frame_time

    def on_exit(self) -> None:
        # Do nothing
        pass

    def get_state_time(self):
        return global_state.frame_time - self.start_time

    def get_frame_index(self) -> int:
        return int(self.get_state_time() * 10)

    def get_frame_index_fast(self) -> int:
        return int(self.get_state_time() * 20)

    def one_shot(self, delta: float, name: str) -> bool:
        if name not in self.events and global_state.frame_time - self.start_time > delta:
            self.events[name] = global_state.frame_time - self.start_time
            return True
        return False

    def every(self, delta: float, name: str, offset: float = 0.0) -> bool:
        if name not in self.events:
            if global_state.frame_time - self.start_time > offset:
                self.events[name] = global_state.frame_time + delta
                return True
        elif global_state.frame_time - self.events[name] >= 0:
            self.events[name] = global_state.frame_time + delta
            return True
        return False

File: game-web/src/state.ts
import { globalState } from './globalState';
import type { GameData } from './gameData';
import type { PhoneEvents } from './phoneEvents';

export type StateClass = new (context: GameData) => State;

export class State {
  startTime: number = 0;
  events: Record<string, number> = {};
  context: GameData;

  constructor(context: GameData) {
    this.context = context;
  }

  processEvents(_phoneEvents: PhoneEvents): StateClass | null {
    return null;
  }

  render(_ctx: OffscreenCanvasRenderingContext2D): void {}

  onEnter(): void {
    this.startTime = globalState.frameTime;
  }

  onExit(): void {}

  getStateTime(): number {
    return globalState.frameTime - this.startTime;
  }

  getFrameIndex(): number {
    return Math.floor(this.getStateTime() * 10);
  }

  getFrameIndexFast(): number {
    return Math.floor(this.getStateTime() * 20);
  }

  oneShot(delta: number, name: string): boolean {
    if (!(name in this.events) && globalState.frameTime - this.startTime > delta) {
      this.events[name] = globalState.frameTime - this.startTime;
      return true;
    }
    return false;
  }

  every(delta: number, name: string, offset: number = 0.0): boolean {
    if (!(name in this.events)) {
      if (globalState.frameTime - this.startTime > offset) {
        this.events[name] = globalState.frameTime + delta;
        return true;
      }
    } else if (globalState.frameTime - this.events[name] >= 0) {
      this.events[name] = globalState.frameTime + delta;
      return true;
    }
    return false;
  }
}
